spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveProcedures.scala

```diff
@@ -0,0 +1,196 @@
+    }
+
+    // verify the number of params in the procedure match the number of params in the method
+    if (params.length != methodType.parameterCount) {
+      throw new AnalysisException("Method parameter count must match the number of procedure parameters")
+    }
+
+    // the MethodHandle API does not allow us to check the generic type
+    // so we only verify the return type is either void or iterable
+
+    if (outputType.nonEmpty && methodReturnType != classOf[java.lang.Iterable[_]]) {
+      throw new AnalysisException(
+        s"Wrong method return type: $methodReturnType; the procedure defines $outputType " +
+        "as its output so must return java.lang.Iterable of Spark Rows")
+    }
+
+    if (outputType.isEmpty && methodReturnType != classOf[Void]) {
+      throw new AnalysisException(
+        s"Wrong method return type: $methodReturnType; the procedure defines no output columns " +
+        "so must be void")
+    }
+  }
+
+  private def buildArgExprs(procedure: Procedure, args: Seq[CallArgument]): Seq[Expression] = {
+    val params = procedure.parameters
+
+    // build a map of declared parameter names to their positions
+    val nameToPositionMap = params.map(_.name).zipWithIndex.toMap
+
+    // build a map of parameter names to args
+    val nameToArgMap = buildNameToArgMap(params, args, nameToPositionMap)
+
+    // verify all required parameters are provided
+    val missingParamNames = params.filter(_.required).collect {
+      case param if !nameToArgMap.contains(param.name) => param.name
+    }
+
+    if (missingParamNames.nonEmpty) {
+      throw new AnalysisException(s"Missing required parameters: ${missingParamNames.mkString("[", ",", "]")}")
+    }
+
+    val argExprs = new Array[Expression](params.size)
+
+    nameToArgMap.foreach { case (name, arg) =>
+      val position = nameToPositionMap(name)
+      val param = params(position)
+      val paramType = param.dataType
+      val argType = arg.expr.dataType
+      if (paramType != argType) {
+        throw new AnalysisException(s"Wrong arg type for ${param.name}: expected $paramType but got $argType")
+      }
+      argExprs(position) = arg.expr
+    }
+
+    // assign nulls to optional params that were not set
+    params.foreach {
+      case p if !p.required && !nameToArgMap.contains(p.name) =>
+        val position = nameToPositionMap(p.name)
+        argExprs(position) = Literal.create(null, p.dataType)
+      case _ =>
+    }
+
+    argExprs
+  }
+
+  private def buildNameToArgMap(
+      params: Seq[ProcedureParameter],
+      args: Seq[CallArgument],
+      nameToPositionMap: Map[String, Int]): Map[String, CallArgument] = {
+
+    val containsNamedArg = args.exists(_.isInstanceOf[NamedArgument])
+    val containsPositionalArg = args.exists(_.isInstanceOf[PositionalArgument])
+
+    if (containsNamedArg && containsPositionalArg) {
+      throw new AnalysisException("Named and positional arguments cannot be mixed")
```

Comment:
This is Presto's current behavior?

Reply:
Yes, Presto does not allow mixing named and positional args.